### PR TITLE
[opengl][refactor] refactor stuff into one "SnodeInfo" for better furture sparse support

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #711 #992

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: taichi/backends/opengl/struct_opengl.cpp [hunk 1]
```diff
@@ -32,41 +31,41 @@ void OpenglStructCompiler::collect_snodes(SNode &snode) {
...
-    stride_map_[node_name] = data_type_size(snode.dt);
+    snode_map_[node_name].stride = data_type_size(snode.dt);
   } else if (snode.type == SNodeType::dense || snode.type == SNodeType::root) {
-    const int n = (snode.type == SNodeType::dense) ? snode.n : 1;
-    length_map_[node_name] = n;
-    stride_map_[node_name] =
-        stride_map_[node_name + "_ch"] * length_map_[node_name];
-    class_children_map_[node_name] = stride_map_.at(node_name + "_ch");
+    const int n = (snode.type == SNodeType::root) ? 1 : snode.n;
+    snode_map_[node_name].length = n;
```
Comment (k-ye) [inline taichi/backends/opengl/struct_opengl.cpp:65]: Similar to what you've done for L38, you can write `auto &snode_info = snode_map_[node_name];` once?

File: taichi/backends/opengl/codegen_opengl.cpp [hunk 2]
```diff
@@ -258,8 +258,8 @@ class KernelGen : public IRVisitor {
   void visit(GetChStmt *stmt) override {
     emit("int {} = {} + {}; // {}", stmt->short_name(),
          stmt->input_ptr->short_name(),
-         struct_compiled_
-             ->class_get_map[stmt->input_snode->node_type_name][stmt->chid],
+         struct_compiled_->snode_map[stmt->input_snode->node_type_name]
```
Comment (k-ye) [inline taichi/backends/opengl/codegen_opengl.cpp:261]: FYI: If you do not want to accidentally insert a default value into the map when the key is missing, consider using [`find`](http://www.cplusplus.com/reference/unordered_map/unordered_map/find/) over `operator[]`.

Comment (archibate) [reply taichi/backends/opengl/codegen_opengl.cpp:261]: I guess [`at`](www.cplusplus.com/reference/map/map/at) should also work?

Comment (k-ye) [reply taichi/backends/opengl/codegen_opengl.cpp]: Yes (sorry i slipped through this PR..)

**Pull request merged**